src/Utils/PathSegment.vala

```diff
@@ -0,0 +1,302 @@
+/**
+ * Copyright (c) 2022 Alecaddd (https://alecaddd.com)
+ *
+ * This file is part of Akira.
+ *
+ * Akira is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+
+ * Akira is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with Akira. If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authored by: Ashish Shevale <shevaleashish@gmail.com>
+*/
+
+/*
+ * Following is a generic structure used for storing all types of segments
+ * is a path.
+ * type : refers to the type of segment.
+ * curve_begin : first point of a curve or end point of line.
+ * tangent_1, tangent_2 : control points for controlling curves.
+ * curve_end : last point of a curve
+ *
+ * We can create a different types of segments using this rep as
+ * 1. Symmetric Bezier Curve (both control points are equi-distant from curve_begin)
+ * 2. Asymmetric Bezier Curve (control points can be at different distances from curve_begin)
+ * 3. Quadratic Curve (single control point)
+ * 4. Line (pretty obvious)
+ */
+public struct Akira.Utils.PathSegment {
+    public Lib.Modes.PathEditMode.Type type;
+
+    // This point is used for drawing lines.
+    // Store point in segment_begin to save space, but use the alias to make code readable.
```

Comment:
I don't understand how you are saving space, you are still copying those values and keeping them in memory--so you have duplicated storage.

Reply:
I also don't think this is part of the fundamental path structure--it should be an API on top.

Reply:
@mbfraga Actually it does save space. 

If you take a look at the generated .c file in *build/src/libakira-library-1.0.a.p/Utils/PathSegment.c*, **PathSegment** is just a struct with a type and 4 points. 

In Vala, any getter and setter variables you create become methods. You can find the method **akira_utils_path_segment_set_line_end** in the .c file.